File: cpp/053_MaximumSubarray.cpp
//dp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        if(nums.empty()) return 0;
        int dp = nums[0];
        int max = dp;
        for(int i=1; i < nums.size(); ++i){
            dp = std::max(dp + nums[i], nums[i]);
            max = std::max(dp,max);
        }
        return max;
    }
};

File: js/053_MaximumSubarray.js
/**
 * @param {number[]} nums
 * @return {number}
 */

//dp
var maxSubArray = function(nums) {
    if(nums.length < 1) return 0;
    var dp = nums[0];
    var max = dp;
    for(var i = 1; i < nums.length; ++i){
        dp = (Math.max(dp + nums[i], nums[i]));
        max = Math.max(max, dp);
    }
    return max;
};

//dc
